streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/TaskMovement.java

```diff
@@ -56,25 +54,34 @@ private int numCaughtUpClients() {
         return caughtUpClients.size();
     }
 
-    private static boolean taskIsNotCaughtUpOnClientAndOtherCaughtUpClientsExist(final TaskId task,
-                                                                                 final UUID client,
-                                                                                 final Map<TaskId, SortedSet<UUID>> tasksToCaughtUpClients) {
-        return !taskIsCaughtUpOnClientOrNoCaughtUpClientsExist(task, client, tasksToCaughtUpClients);
+    private static boolean taskIsNotCaughtUpOnClientAndOtherMoreCaughtUpClientsExist(final TaskId task,
+                                                                                     final UUID client,
+                                                                                     final Map<UUID, ClientState> clientStates,
+                                                                                     final Map<TaskId, SortedSet<UUID>> tasksToCaughtUpClients,
+                                                                                     final Map<TaskId, List<UUID>> tasksToClientByLag) {
```

Comment:
Sure I've pushed up a commit that does this.
My only concern is that these `SortedSet`'s now end up holding a reference to `clientStates` rather than just being a plain old list/set etc.